### ISSUE TITLE
Please make it possible to create/link a file from a program without having to emulate browser

### ISSUE THREAD

**Issue by ssteinerx** [marvinlabs/customer-area#74]
As you know, I'm writing an addon that allows serving of client files from Amazon S3.

I would like a function that:

  1. Creates a post
  2. Attaches a file to the post
  3. Triggers appropriate notifications

Having to emulate a logged in user to perform this collection of functions is just way too much work, nor is it documented making any code that relies on specific calls within the plugin(s) very fragile vis a vis updates to the plugin.

Thanks,

Steve
ssteinerX

**Comment by vpratfr** [comment]
For the file, would passing a local path to the server be OK ?

**Comment by ssteinerx** [comment]
That would be OK.  The way I'm using it is by running a cron job that checks for new files and, when it finds them, moves them to a sync folder (e.g. sync/Client Group Name/new_file).  I've actually got the S3 mounted as a file system, but safe mode prevents direct access so I move the file somewhere PHP can get at it.  When there's something new in the sync folder, I have to look up "Client Group Name" in the database, make the post, lookup the directory etc.  A documented way of getting the group's ID, then passing that to the new function we're discussing would be fine.

**Comment by ssteinerx** [comment]
Also, there are a zillion filters that trigger (notifications etc.) so it would be important to have all that happen automatically as well.  Since it's not documented, it's very hard to figure out all of the things to make all the other plugins do their various things...

**Comment by vpratfr** [comment]
Yes of course, all filters will also be executed. In fact that function should only be using the existing core add-ons so no worries on that side I think.

**Comment by ssteinerx** [comment]
Should I stop writing my own in anticipation of this being released some time soon?  I have to deliver this functionality to the client and I'm sure I'll not get the notification part (at least!) right.  Then I can go work on the S3 side of things while I wait...

Thanks,

Steve

**Comment by vpratfr** [comment]
I'll try to get this function in 6.1. shoud be released by April end. but you'll get early access to it in the develop branch when I close this branch.

By the way, there are some implementations for S3 in Easy Digital Downloads add-ons, you may find it interesting to lok at how they handle it. This could give you ideas.

**Comment by vpratfr** [comment]
I was doing it and I tought that maybe we are doing it wrong.

You are having all files on an external storage usually because you don't want to clutter the space on your web server.

Hence, if you are simply synchronizing files from S3 onto the web server, we are basically loosing all advantage of using S3. And same would happen if you were using dropbox instead of S3.

In my opinion, the right way to do it would be:

- just like we have an upload method "from FTP folder", we should be able to select something like "from AmazonS3 folder" 
- this would populate a select/list box showing all files and folders in that remote place
- when we are serving the file to a user downloading it, stream it from S3 directly

What do you think?

**Comment by vpratfr** [comment]
Here is a start: 8f49dd8115f7be7573e65430baedd7bae8d402bd

**Comment by ssteinerx** [comment]
Thanks!  I'm going to pull this right into my site and beat it around.  Expect a pull request by the end of the day (my time GMT -4) to fix anything I run into.

Thanks again,
Steve

**Comment by vpratfr** [comment]
See also #76 I think we should do it right (not with clumsy cron jobs and full file copy on both S3 and the web server)

**Comment by ssteinerx** [comment]
I'm not sure if you know, but the way WordPress "cron" works is by checking for "cron jobs" whose time has passed when anyone visits the site.  You can make this happen on a schedule by running a real cron job that just hits any part of the site with curl or whatever.

In my particular case, I have root on the server, so I'm using s3fs to mount the S3 bucket as a file system, then  pulling new files over into the FTP area with a root-level cron job.  They're then picked up by my auto-sync plugin and then are automatically pushed to the appropriate group's folder by creating a post etc.  

There are a few difficulties with doing this right off S3:
  1> Connecting to and scanning S3 can take longer than PHP has to run
  2> Because of the nature of server side scripting, you can't just thread it out in the background -- it's all synchronous.

  If this was in node.js, for example, you'd just lazily scan the S3 in a separate process or use fsevents and such to be notified of changes (on a mounted volume with s3fs).

  At the moment, I've got to get this working for the client, so I'm not shooting for optimum, nor is any of the client-specific code going to be entangled with any work I do on your code.  On your side, I'm just wanting the post to be created, the FTP file to be properly moved, and the proper hooks to all fire.

Steve

**Comment by vpratfr** [comment]
I know how CRON works in WP. and this is exactly why I'd like to avoid it ;)

Regarding your concerns, PHP execution time is usually set to 30 seconds. If S3 does not connect in 30 seconds, that usually means there is a problem. For a directory listing, I would assume the same, there can also be instructions to the user in order to avoid putting all his files into the same folder, which would reduce the time to scan the S3 folder.

And to get some async goodness, we can always throw in some Ajax.

**Comment by ssteinerx** [comment]
Agreed on the WP cron ;)

It's not the connecting time I'm worried about, it's the scanning time.  This particular use case involves a client posting monthly reports and the entire history for all clients has to be available.  I.E. Clients * Years * 12 * 2 reports/month.  Quite a lot of stuff.  I had originally wanted to just use s3fs, but it has some issues where excessively long scans can actually consume all CPU, leaving the server essentially hung.  I may actually write my sync stuff in node.js and run it on a local machine to avoid this issue entirely.

**Comment by vpratfr** [comment]
How about storing those files inside a folder structure like `/s3storage/client/year/month/` and on the client side, just query with Ajax the top folder.

Hence, this would go like:

- I add a private file
- I select S3 from the file upload methods (Ajax builds a select with a list of clients)
- I select the client (ajax builds a select with a list of years)
- I select the year (ajax builds a select with a list of months)
- I select the month (ajax builds a select with a list of files)
- I select the file and hit publish (the S3 file path is associated to the private file)

Of course, this looks pretty tedious. Maybe you could combine year-month into a single select

The good thing about this is that this will be generic and actually usable by anyone so your addon will sell better.

**Comment by ssteinerx** [comment]
This looks good (if tedious) for the plugin version, but the whole point, for my current use case, is to eliminate the need to interact via the web interface completely.

I also noticed that you limit the call to 1 file per call, then loop over the files array.  Is that restriction just for testing, or is there a good reason you can't add more than one file at a time?

**Comment by vpratfr** [comment]
Currently we can only attach a single file to a private file post. However, I want to remove this limitation sooner or later, hence I have made the function ready for it even if it curently does not use the array.

**Comment by vpratfr** [comment]
@ssteinerx I am currently working on #31 and even though this is not directly related, I would love if you could give some feedback about the branch [feature/admin-cleanup](https://github.com/marvinlabs/customer-area/tree/feature/admin-cleanup) (try it without any add-on enabled, there are some changes that require updating them). 

Could you give some input about that in #31?

**Comment by ssteinerx** [comment]
I'm working on it again today...spent yesterday disinfecting a client site.

S

**Comment by vpratfr** [comment]
Closing this as I have merged the functions which are supposed to allow someone to do it. Open a new issue if anything is missing.
